### PR TITLE
First proofs for keyring_trace module

### PR DESCRIPTION
**Pull request by tegansiobhan**
Harnesses for proofs for keyring aws_cryptosdk_keyring_trace_init, aws_cryptosdk_keyring_trace_record_clean_up, aws_cryptosdk_keyring_trace_record_init_clone and aws_cryptosdk_string_dup (called from the record_init_clone function).

Added corresponding preconditions into the source code for keyring_trace.c and utils.c. 

Also, updated Makefile.common to include the UNWINDSET of CBMC (already done in PR 395 but that hasn't been merged yet) and created Makefile.string to specify the MAX_STRING_LEN used in proofs. 

@danielsn @bdonlan @ttjsu-aws

### CODE REVIEW

**Comment by tegansiobhan** [comment]
@danielsn @dougch

File: .cbmc-batch/jobs/Makefile.common [hunk 1]
```diff
@@ -307,4 +307,4 @@ gitclean: veryclean
 
 .PHONY: setup_dependencies cbmc property coverage report clean veryclean
 
-include $(HELPERDIR)/jobs/Makefile.cbmc_batch
+include $(HELPERDIR)/jobs/Makefile.cbmc_batch
```
Comment (danielsn) [inline .cbmc-batch/jobs/Makefile.common]: add a newline

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_init/Makefile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########
+# if Makefile.local exists, use it. This provides a way to override the defaults 
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+UNWINDSET +=
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_init/Makefile]: I've started sorting these sections alphabetically.  Makes the makefile easier to read.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_init/aws_cryptosdk_keyring_trace_init_harness.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    struct aws_allocator *alloc = can_fail_allocator(); /* Precondition: alloc must be non-null */
+    struct aws_array_list trace;                        /* Precondition: trace must be non-null */
+
+    if (aws_cryptosdk_keyring_trace_init(alloc, &trace) == AWS_OP_SUCCESS) {
+        /* assertions */
+        assert(aws_array_list_is_valid(&trace));
+        assert(trace.alloc == alloc);
+        assert(trace.item_size == sizeof(struct aws_cryptosdk_keyring_trace_record));
+        assert(trace.length == 0);
+    }
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_init/aws_cryptosdk_keyring_trace_init_harness.c]: Is there anything to assert on failure (no is an acceptable answer).

File: .cbmc-batch/jobs/aws_cryptosdk_serialize_frame/aws_cryptosdk_serialize_frame_harness.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    /* Save the old state of the ciphertext buffer */
+    uint8_t *old_ciphertext_buffer   = ciphertext_buf.buffer;
+    size_t old_ciphertext_buffer_len = ciphertext_buf.len;
+
+    int rval = aws_cryptosdk_serialize_frame(&frame, &ciphertext_size, plaintext_size, &ciphertext_buf, &alg_props);
+    if (rval == AWS_OP_SUCCESS) {
+        assert(aws_cryptosdk_frame_is_valid(&frame));
+        assert(ciphertext_buf.buffer == old_ciphertext_buffer);
+        assert(ciphertext_buf.len == old_ciphertext_buffer_len + ciphertext_size);
+    } else {
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_serialize_frame/aws_cryptosdk_serialize_frame_harness.c]: Nice

File: .cbmc-batch/source/make_common_data_structures.c [hunk 5]
```diff
@@ -21,6 +21,15 @@
 #include <evp_utils.h>
 #include <make_common_data_structures.h>
 
+void ensure_alg_properties_has_allocated_names(struct aws_cryptosdk_alg_properties *const alg_props) {
+    size_t md_name_size;
+    alg_props->md_name = can_fail_malloc(md_name_size);
```
Comment (danielsn) [inline .cbmc-batch/source/make_common_data_structures.c]: this is actually testing somewhat more cases than the function name suggests.  You either allocate, or have null.

File: include/aws/cryptosdk/cipher.h [hunk 6]
```diff
@@ -101,6 +101,12 @@ struct aws_cryptosdk_alg_properties {
 AWS_CRYPTOSDK_API
 const struct aws_cryptosdk_alg_properties *aws_cryptosdk_alg_props(enum aws_cryptosdk_alg_id alg_id);
 
+/**
+ * Checks whether an aws_cryptosdk_alg_properties is valid and is supported by the SDK.
```
Comment (danielsn) [inline include/aws/cryptosdk/cipher.h]: Maybe mention the runtime (constant, I assume)

File: include/aws/cryptosdk/private/framefmt.h [hunk 7]
```diff
@@ -33,6 +34,20 @@ struct aws_cryptosdk_frame {
     struct aws_byte_buf authtag;
 };
 
+// MAX_FRAME_SIZE = 2^32 - 1
+#define MAX_FRAME_SIZE 0xFFFFFFFF
```
Comment (danielsn) [inline include/aws/cryptosdk/private/framefmt.h]: UINT32_MAX

File: include/aws/cryptosdk/private/framefmt.h [hunk 8]
```diff
@@ -33,6 +34,20 @@ struct aws_cryptosdk_frame {
...
+// MAX_FRAMES = 2^32 - 1
+#define MAX_FRAMES 0xFFFFFFFF
+// MAX_UNFRAMED_PLAINTEXT_SIZE = 2^36 - 32
+#define MAX_UNFRAMED_PLAINTEXT_SIZE 0xFFFFFFFE0ull
+
+/**
+ * Checks whether a frame struct is valid. At the moment this means
+ * that it checks the validity of the byte buffers and the fact that
+ * they should have NULL allocators.
+ */
```
Comment (danielsn) [inline include/aws/cryptosdk/private/framefmt.h]: runtime

File: source/cipher.c [hunk 9]
```diff
@@ -88,6 +88,32 @@ static enum aws_cryptosdk_sha_version aws_cryptosdk_which_sha(enum aws_cryptosdk
     }
 }
 
+static bool alg_properties_equal(
```
Comment (danielsn) [inline source/cipher.c]: should this be `aws_cryptosdk_alg_properties_equal`

File: source/framefmt.c [hunk 10]
```diff
@@ -235,16 +254,26 @@ int aws_cryptosdk_serialize_frame(
     size_t plaintext_size,
     struct aws_byte_buf *ciphertext_buf,
     const struct aws_cryptosdk_alg_properties *alg_props) {
+    AWS_PRECONDITION(aws_cryptosdk_alg_properties_is_valid(alg_props));
     struct aws_cryptosdk_framestate state;
 
+    // The plaintext_size should be bound to prevent arithmetic
+    // overflows due to addition
+    if ((frame->type == FRAME_TYPE_SINGLE && plaintext_size > MAX_UNFRAMED_PLAINTEXT_SIZE) ||
```
Comment (danielsn) [inline source/framefmt.c]: Are these bounds documented in the documentation @ttjsu-aws

File: source/utils.c [hunk 11]
```diff
@@ -38,6 +39,9 @@ int aws_cryptosdk_hash_elems_array_init(
 }
 
 struct aws_string *aws_cryptosdk_string_dup(struct aws_allocator *alloc, const struct aws_string *str) {
-    if (str->allocator) return aws_string_new_from_string(alloc, str);
+    AWS_FATAL_PRECONDITION(alloc != NULL);
```
Comment (danielsn) [inline source/utils.c]: can now use the `aws_allocator_is_valid()` function from c common

File: .cbmc-batch/jobs/Makefile.string [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+# if Makefile.local exists, use it. This provides a way to override the defaults 
+sinclude ../Makefile.local
+#otherwise, use the default values
+include ../Makefile.local_default
+
+# Sufficently long to get 100% coverage on the string APIs
+# short enough that all proofs complete quickly
+MAX_STRING_LEN ?= 16
+
+DEFINES += -DMAX_STRING_LEN=$(MAX_STRING_LEN)
```
Comment (ttjsu-aws) [inline .cbmc-batch/jobs/Makefile.string]: A new line for end of file is required.

File: .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_clean_up/aws_cryptosdk_keyring_trace_record_clean_up_harness.c [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/cryptosdk/private/keyring_trace.h>
+#include <make_common_data_structures.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_cryptosdk_keyring_trace_record_clean_up_harness() {
+    /* data structure */
+    struct aws_cryptosdk_keyring_trace_record record; /* Precondition: record is non-null */
```
Comment (ttjsu-aws) [inline .cbmc-batch/jobs/aws_cryptosdk_keyring_trace_record_clean_up/aws_cryptosdk_keyring_trace_record_clean_up_harness.c]: Can we assert that the record is non-null before we call the clean up function?

File: source/framefmt.c (hunk 10 above)
Comment (ttjsu-aws) [reply source/framefmt.c]: Yes, we currently have that: 

- The total bytes allowed in a single frame MUST be less than or equal to 2^32 - 1.
- The number of frames in a single message MUST be less than or equal to 2^32 - 1.
- The encrypted content length MUST NOT be greater than 2^36 - 32, or 64 gibibytes (64 GiB).

File: source/keyring_trace.c [hunk 14]
```diff
@@ -16,10 +16,14 @@
 #include <aws/cryptosdk/private/utils.h>
 
 void aws_cryptosdk_keyring_trace_record_clean_up(struct aws_cryptosdk_keyring_trace_record *record) {
+    AWS_FATAL_PRECONDITION(record != NULL);
```
Comment (danielsn) [inline source/keyring_trace.c]: `AWS_OBJECT_PTR_IS_VALID`

File: source/keyring_trace.c [hunk 15]
```diff
@@ -115,21 +118,44 @@ int aws_cryptosdk_keyring_trace_add_record_buf(
 
 int aws_cryptosdk_keyring_trace_init(struct aws_allocator *alloc, struct aws_array_list *trace) {
     // arbitrary starting point, list will resize as necessary
+    AWS_FATAL_PRECONDITION(trace != NULL);
```
Comment (danielsn) [inline source/keyring_trace.c]: `AWS_OBJECT_PTR_IS_VALID`

File: source/keyring_trace.c [hunk 16]
```diff
@@ -115,21 +118,44 @@ int aws_cryptosdk_keyring_trace_add_record_buf(
...
+        AWS_POSTCONDITION(trace->length == 0);
+    }
+    return r_val;
 }
 
 int aws_cryptosdk_keyring_trace_record_init_clone(
     struct aws_allocator *alloc,
     struct aws_cryptosdk_keyring_trace_record *dest,
     const struct aws_cryptosdk_keyring_trace_record *src) {
+    AWS_FATAL_PRECONDITION(alloc != NULL);
```
Comment (danielsn) [inline source/keyring_trace.c]: `aws_allocator_is_valid()`

File: source/keyring_trace.c [hunk 17]
```diff
@@ -115,21 +118,44 @@ int aws_cryptosdk_keyring_trace_add_record_buf(
...
+    AWS_FATAL_PRECONDITION(dest != NULL);
+    AWS_FATAL_PRECONDITION(src != NULL);
```
Comment (danielsn) [inline source/keyring_trace.c]: We ought to be able to assert src is valid

**Comment by tawdry-audrey** [comment]
Rebased, updated, and merged in https://github.com/aws/aws-encryption-sdk-c/pull/482.

**Pull request closed without merging**